x-pack/plugins/monitoring/server/deprecations.ts

```diff
@@ -16,8 +16,32 @@ import { CLUSTER_ALERTS_ADDRESS_CONFIG_KEY } from '../common/constants';
  * major version!
  * @return {Array} array of rename operations and callback function for rename logging
  */
-export const deprecations = ({ rename }: ConfigDeprecationFactory): ConfigDeprecation[] => {
+export const deprecations = ({
+  rename,
+  renameFromRoot,
+}: ConfigDeprecationFactory): ConfigDeprecation[] => {
   return [
+    renameFromRoot('xpack.monitoring', 'monitoring'),
+    renameFromRoot('xpack.monitoring.max_bucket_size', 'monitoring.ui.max_bucket_size'),
+    renameFromRoot('xpack.monitoring.min_interval_seconds', 'monitoring.ui.min_interval_seconds'),
+    renameFromRoot(
+      'xpack.monitoring.show_license_expiration',
+      'monitoring.ui.show_license_expiration'
+    ),
+    renameFromRoot(
+      'xpack.monitoring.ui.container.elasticsearch.enabled',
+      'monitoring.ui.container.elasticsearch.enabled'
+    ),
+    renameFromRoot(
+      'xpack.monitoring.ui.container.logstash.enabled',
+      'monitoring.ui.container.logstash.enabled'
+    ),
+    renameFromRoot('xpack.monitoring.elasticsearch', 'monitoring.ui.elasticsearch'),
```

Comment:
@restrry This blanket rename seems to work for all sub keys - is this the recommended way?

Reply:
It does work. Deprecations are supporting rename/delete of non-final level keys, meaning you can rename a nested group of keys in a single operation when the nested structure doesn't change.

However it seems it's not the case for you as with: 

```ts
renameFromRoot('xpack.monitoring', 'monitoring'),
renameFromRoot('xpack.monitoring.max_bucket_size', 'monitoring.ui.max_bucket_size'),
```

You are going to have the `xpack.monitoring.max_bucket_size` key copied in both `monitoring.max_bucket_size` and `monitoring.ui.max_bucket_size` which I believe is not the expected output?

As deprecations are executed in order, you could, technically, have something like

```ts
renameFromRoot('xpack.monitoring', 'monitoring'),
rename('max_bucket_size', 'ui.max_bucket_size'),
```

It would work, the issue is that if the `xpack.monitoring.max_bucket_size' is present, it will output a warning with the wrong path:

`"monitoring.max_bucket_size" is deprecated and has been replaced by "monitoring.ui.max_bucket_size"`

So I think you will have to get back to the state the deprecations were in `core_deprecations.ts`.

You can probably factorise this a little though with something like (very quick example)

```ts
export const deprecations : ConfigDeprecationProvider = ({rename, renameFromRoot}) => {
    const rootKeys = ['kibana.collection.enabled', ...];
    const uiKeys = ['ccs.enabled', ...]
   return [
      ...rootKeys.map(key => renameFromRoot(`xpack.monitoring.${key}`, `monitoring.${key}`),
     ...uiKeys.map(key => renameFromRoot(`xpack.monitoring.${key}`, `monitoring.ui.${key}`),
   ]
}
```

Another option would be to implement you own version of `ConfigDeprecation`, to have more control on own you mutate the data instead of just using the `rename/renameFronRoot` builtins. You can check `src/core/server/config/deprecation/deprecation_factory.ts` and `src/core/server/config/deprecation/core_deprecations.ts` for some examples.

Reply:
What if I just switch the order? So the blanket rename happens last? With that and this config:
```
xpack.monitoring.max_bucket_size: 5000
xpack.monitoring.min_interval_seconds: 9
```

I see:

```
server    log   [11:37:17.142] [warning][config][deprecation] "xpack.monitoring.max_bucket_size" is deprecated and has been replaced by "monitoring.ui.max_bucket_size"
server    log   [11:37:17.143] [warning][config][deprecation] "xpack.monitoring.min_interval_seconds" is deprecated and has been replaced by "monitoring.ui.min_interval_seconds"
server    log   [11:37:17.143] [warning][config][deprecation] "xpack.monitoring" is deprecated and has been replaced by "monitoring". However both key are present, ignoring "xpack.monitoring"
```

Reply:
@chrisronline doesn't this make it order dependable then? But, hopefully we won't need to touch this code again until we remove the warnings altogether, so I think it's fine for now.

However, if we do end up revisiting this, I really like the factory approach @pgayvallet mentioned ++

Reply:
> What if I just switch the order? So the blanket rename happens last?

It would work and I'm not opposed to that, even if as @igoristic said, it would make the deprecations behavior depending on their order, so that's your call.

Reply:
I think it's fine for now. We won't need to make changes to it and we'll remove it in 8.0